crates/turbopack-ecmascript/src/chunk/mod.rs

```diff
@@ -625,11 +629,42 @@ impl EcmascriptChunkContentVc {
         }
         code += "]);\n";
         if this.evaluate.is_some() {
-            let runtime_code = embed_file!("js/src/runtime.js").await?;
-            match &*runtime_code {
-                FileContent::NotFound => return Err(anyhow!("runtime code is not found")),
+            // When a chunk is executed, it will either register itself with the current
+            // instance of the runtime, or it will push itself onto the list of pending
+            // chunks (`self.TURBOPACK`).
+            //
+            // When the runtime executes, it will pick up and register all pending chunks,
+            // and replace the list of pending chunks with itself so later chunks can
+            // register directly with it.
+            code += indoc! { r#"
+                (() => {
+                if (!Array.isArray(globalThis.TURBOPACK)) {
+                    return;
+                }
+            "# };
+
+            let specific_runtime_code = match *this.environment.chunk_loading().await? {
+                ChunkLoading::None => return Err(anyhow!("unsupported environment")),
+                ChunkLoading::NodeJs => embed_file!("js/src/runtime.nodejs.js").await?,
+                ChunkLoading::Dom => embed_file!("js/src/runtime.dom.js").await?,
+            };
+
+            match &*specific_runtime_code {
+                FileContent::NotFound => return Err(anyhow!("specific runtime code is not found")),
                 FileContent::Content(file) => code.push_source(file.content(), None),
             };
+
+            let shared_runtime_code = embed_file!("js/src/runtime.js").await?;
+
+            match &*shared_runtime_code {
+                FileContent::NotFound => return Err(anyhow!("shared runtime code is not found")),
+                FileContent::Content(file) => code.push_source(file.content(), None),
+            };
+
+            code += indoc! { r#"
+                globalThis.TURBOPACK = { push: registerChunk };
```

Comment:
Why is this line no longer in `runtime.js`?

Reply:
It probably should be, fixing now.

Reply:
This was to circumvent a typing error where pulling out the array check above makes TS unhappy about the type of `globalThis.TURBOPACK`. I'll see if I can figure out a better way.